### PR TITLE
Fix ruff rule F401: remove unused imports

### PR DESCRIPTION
**Pull request by connortann**
This PR removes unused imports, supporting #21.

This change leads to a large number of changed files, so I think it would be prudent to do a thorough code review to ensure nothing is accidentally removed from the package's public API.

There were a couple of cases where unused imports seem to be deliberate, which I'll summarise here if that helps with code review:

### Package imports

Some unused imports in `__init__.py` modules are deliberate to expose the objects in that namespace, e.g.:

https://github.com/dsgibbons/shap/blob/a0e5c00ddcdc7ca0e3fd0a5f22876e5438e8ecc8/shap/maskers/__init__.py#L1-L4

So, I propose excluding those files from this check with:

```python
[tool.ruff.per-file-ignores]
# Unused imports often intentional in top-level __init__.py 
"__init__.py" = ["F401"]
```

### Isinstance checks

In other cases, optional dependencies are imported in this manner:

https://github.com/dsgibbons/shap/blob/a0e5c00ddcdc7ca0e3fd0a5f22876e5438e8ecc8/shap/maskers/_image.py#L8-L11

And then used with `isinstance` checks, in a manner that won't fail if the package in question is not installed:

https://github.com/dsgibbons/shap/blob/a0e5c00ddcdc7ca0e3fd0a5f22876e5438e8ecc8/shap/maskers/_image.py#L76

So, I've adedd `# noqa` comments to those cases.

### CODE REVIEW

File: shap/explainers/_tree.py [hunk 1]
```diff
@@ -818,7 +814,6 @@ def __init__(self, model, data=None, data_missing=None, model_output=None):
             else:
                 assert False, "Unsupported Spark model type: " + str(type(model))
         elif safe_isinstance(model, "xgboost.core.Booster"):
-            import xgboost
```
Comment (connortann) [inline shap/explainers/_tree.py:821]: Unsure if this import has a runtime effect?

**Comment by connortann** [comment]
Related: https://github.com/slundberg/shap/pull/2595

**Comment by dsgibbons** [comment]
The imports have somehow affected the floating point precision for Python 3.8. I have heard about this somewhere before, but I can't immediately find anything online. Something about TensorFlow or some related package affecting floating point precision upon import. Perhaps we just provide a larger tolerance for the failing test?

**Comment by dsgibbons** [comment]
There seems to be some randomness around floating-point precision. Upon merging #96, I had a [similar test fail](https://github.com/dsgibbons/shap/actions/runs/5135384415/attempts/1). I reran GitHub Actions and [everything passed again](https://github.com/dsgibbons/shap/actions/runs/5135384415). I'm rerunning the failing test here to see if that fixes anything.

**Comment by dsgibbons** [comment]
Passing now. I think we should merge this PR and raise a new issue around floating-point precision and GitHub Actions.

**Review by dsgibbons — APPROVED** [review]
Approved, but I'll let @connortann merge.

**Comment by dsgibbons** [comment]
Raised #98 to investigate floating-point precision issues.

**Pull request merged**